src/test/java/plugins/OicAuthPluginTest.java

```diff
@@ -0,0 +1,254 @@
+        securityRealm.setAutomaticConfiguration(String.format("%s/realms/%s/.well-known/openid-configuration", keycloakUrl, REALM));
+        securityRealm.logoutFromOpenidProvider(true);
+        securityRealm.setPostLogoutUrl(jenkins.url("OicLogout").toExternalForm());
+        securityRealm.setUserFields(null, null, null, "groups");
+        sc.useAuthorizationStrategy(LoggedInAuthorizationStrategy.class);
+        sc.save();
+    }
+
+    @Test
+    public void fromJenkinsToKeycloak() {
+        final KeycloakUtils.User bob = new KeycloakUtils.User(userBobKeycloakId, "bob", "bob@acme.org", "Bob", "Smith");
+        final KeycloakUtils.User john = new KeycloakUtils.User(userJohnKeycloakId, "john", "john@acme.org", "John", "Smith");
+        jenkins.open();
+
+        jenkins.clickLink("log in");
+        keycloakUtils.login(bob.getUserName());
+        assertLoggedUser(bob, "jenkinsAdmin");
+
+        jenkins.logout();
+        jenkins.open();
+        assertLoggedOut();
+
+        // logout from Jenkins does mean logout from keycloak
+        jenkins.open();
+
+        clickLink("log in");
+        keycloakUtils.login(john.getUserName());
+        assertLoggedUser(john, "jenkinsRead");
+    }
+
+    @Test
+    public void fromKeycloakToJenkins() throws Exception {
+        final KeycloakUtils.User bob = new KeycloakUtils.User(userBobKeycloakId, "bob", "bob@acme.org", "Bob", "Smith");
+        final KeycloakUtils.User john = new KeycloakUtils.User(userJohnKeycloakId, "john", "john@acme.org", "John", "Smith");
+        final String loginUrl = String.format("%s/realms/%s/account", keycloak.getAuthServerUrl(), REALM);
+        keycloakUtils.open(new URL(loginUrl));
+
+        keycloakUtils.login(bob.getUserName());
+        jenkins.open();
+        jenkins.clickLink("log in"); // won't request a login, but log in directly with user from
+
+        assertLoggedUser(bob, "jenkinsAdmin");
+
+        keycloakUtils.logout(bob);
+        jenkins.open();
+        jenkins.logout(); // logout from keycloak does not logout from Jenkins (seems not supported in the plugin)
+        assertLoggedOut();
+
+        // Once logged out, we can change the user
+        jenkins.open();
+        jenkins.clickLink("log in");
+        keycloakUtils.login(john.getUserName());
+        assertLoggedUser(john, "jenkinsRead");
+    }
+
+    private void assertLoggedOut() {
+        assertNull("User has logged out from Jenkins", jenkins.getCurrentUser().id());
+
+        assertThrows("User has logged out from keycloak", NoSuchElementException.class,
+                     () -> keycloakUtils.getUser(keycloak.getAuthServerUrl(), REALM));
+    }
+
+    private void assertLoggedUser(KeycloakUtils.User expectedUser, String roleToCheck) {
+        assertThat("User has logged in Jenkins", jenkins.getCurrentUser().id(), is(expectedUser.getId()));
+        jenkins.visit("whoAmI");
+
+        // TODO if needed for more tests, consider to create a proper WhoAmI PageObject
```

Comment:
Use the `API` luke!
`$jenkinsUrl/whoAmI/api/xml?tree=authorities`  (or use `getJson` but that would require a `PageObject` (at least an anonymous inner class) 

Then any page redesign will not break the test!

Reply:
That was my first try and the api is not so simple to use here

**whoAmI/api/json** -> generated output differs from Chrome and Firefox
`jenkins.getPageSource()` will return in firefox

```
<html platform="linux" class="theme-dark" dir="ltr" style="scroll-behavior: initial;"><head><meta http-equiv="Content-Security-Policy" content="default-src 'none'; script-src resource:; img-src 'self';"><link rel="stylesheet" type="text/css" href="chrome://devtools-jsonview-styles/content/main.css"><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="viewer-config" src="resource://devtools/client/jsonview/viewer-config.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="json-viewer" src="resource://devtools/client/jsonview/json-viewer.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/vendor/react-dom" src="resource://devtools/client/shared/vendor/react-dom.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/react-utils" src="resource://devtools/client/shared/react-utils.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/jsonview/components/MainTabbedArea" src="resource://devtools/client/jsonview/components/MainTabbedArea.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/tree/TreeView" src="resource://devtools/client/shared/components/tree/TreeView.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/vendor/react" src="resource://devtools/client/shared/vendor/react.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/vendor/react-prop-types" src="resource://devtools/client/shared/vendor/react-prop-types.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/jsonview/components/JsonPanel" src="resource://devtools/client/jsonview/components/JsonPanel.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/jsonview/components/TextPanel" src="resource://devtools/client/jsonview/components/TextPanel.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/jsonview/components/HeadersPanel" src="resource://devtools/client/jsonview/components/HeadersPanel.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/tabs/Tabs" src="resource://devtools/client/shared/components/tabs/Tabs.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/vendor/react-dom-factories" src="resource://devtools/client/shared/vendor/react-dom-factories.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/jsonview/components/JsonToolbar" src="resource://devtools/client/jsonview/components/JsonToolbar.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/constants" src="resource://devtools/client/shared/components/reps/reps/constants.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/rep" src="resource://devtools/client/shared/components/reps/reps/rep.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/jsonview/components/TextToolbar" src="resource://devtools/client/jsonview/components/TextToolbar.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/jsonview/components/LiveText" src="resource://devtools/client/jsonview/components/LiveText.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/jsonview/components/Headers" src="resource://devtools/client/jsonview/components/Headers.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/jsonview/components/HeadersToolbar" src="resource://devtools/client/jsonview/components/HeadersToolbar.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/tree/ObjectProvider" src="resource://devtools/client/shared/components/tree/ObjectProvider.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/tree/TreeRow" src="resource://devtools/client/shared/components/tree/TreeRow.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/tree/TreeHeader" src="resource://devtools/client/shared/components/tree/TreeHeader.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/scroll" src="resource://devtools/client/shared/scroll.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/jsonview/components/SearchBox" src="resource://devtools/client/jsonview/components/SearchBox.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/jsonview/components/reps/Toolbar" src="resource://devtools/client/jsonview/components/reps/Toolbar.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/undefined" src="resource://devtools/client/shared/components/reps/reps/undefined.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/null" src="resource://devtools/client/shared/components/reps/reps/null.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/string" src="resource://devtools/client/shared/components/reps/reps/string.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/number" src="resource://devtools/client/shared/components/reps/reps/number.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/array" src="resource://devtools/client/shared/components/reps/reps/array.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/object" src="resource://devtools/client/shared/components/reps/reps/object.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/symbol" src="resource://devtools/client/shared/components/reps/reps/symbol.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/infinity" src="resource://devtools/client/shared/components/reps/reps/infinity.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/nan" src="resource://devtools/client/shared/components/reps/reps/nan.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/accessor" src="resource://devtools/client/shared/components/reps/reps/accessor.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/accessible" src="resource://devtools/client/shared/components/reps/reps/accessible.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/attribute" src="resource://devtools/client/shared/components/reps/reps/attribute.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/big-int" src="resource://devtools/client/shared/components/reps/reps/big-int.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/date-time" src="resource://devtools/client/shared/components/reps/reps/date-time.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/document" src="resource://devtools/client/shared/components/reps/reps/document.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/document-type" src="resource://devtools/client/shared/components/reps/reps/document-type.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/event" src="resource://devtools/client/shared/components/reps/reps/event.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/function" src="resource://devtools/client/shared/components/reps/reps/function.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/promise" src="resource://devtools/client/shared/components/reps/reps/promise.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/regexp" src="resource://devtools/client/shared/components/reps/reps/regexp.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/stylesheet" src="resource://devtools/client/shared/components/reps/reps/stylesheet.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/comment-node" src="resource://devtools/client/shared/components/reps/reps/comment-node.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/element-node" src="resource://devtools/client/shared/components/reps/reps/element-node.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/text-node" src="resource://devtools/client/shared/components/reps/reps/text-node.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/error" src="resource://devtools/client/shared/components/reps/reps/error.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/window" src="resource://devtools/client/shared/components/reps/reps/window.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/object-with-text" src="resource://devtools/client/shared/components/reps/reps/object-with-text.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/object-with-url" src="resource://devtools/client/shared/components/reps/reps/object-with-url.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/grip-array" src="resource://devtools/client/shared/components/reps/reps/grip-array.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/grip-entry" src="resource://devtools/client/shared/components/reps/reps/grip-entry.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/grip-map" src="resource://devtools/client/shared/components/reps/reps/grip-map.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/grip" src="resource://devtools/client/shared/components/reps/reps/grip.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/tree/TreeCell" src="resource://devtools/client/shared/components/tree/TreeCell.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/tree/LabelCell" src="resource://devtools/client/shared/components/tree/LabelCell.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/focus" src="resource://devtools/client/shared/focus.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/rep-utils" src="resource://devtools/client/shared/components/reps/reps/rep-utils.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/shared/dom-node-constants" src="resource://devtools/client/shared/components/reps/shared/dom-node-constants.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/reps/prop-rep" src="resource://devtools/client/shared/components/reps/reps/prop-rep.js"></script><script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="devtools/client/shared/components/reps/shared/grip-length-bubble" src="resource://devtools/client/shared/components/reps/shared/grip-length-bubble.js"></script></head><body><div id="content"><div class="tabs tabs-tall "><nav class="tabs-navigation"><ul class="tabs-menu" role="tablist"><li class="tabs-menu-item json is-active" role="presentation"><span class="devtools-tab-line"></span><a id="json-tab" tabindex="0" title="JSON" aria-controls="json-panel" aria-selected="true" role="tab" data-tab-index="0">JSON</a></li><li class="tabs-menu-item rawdata " role="presentation"><span class="devtools-tab-line"></span><a id="rawdata-tab" tabindex="-1" title="Raw Data" aria-controls="rawdata-panel" aria-selected="false" role="tab" data-tab-index="1">Raw Data</a></li><li class="tabs-menu-item headers " role="presentation"><span class="devtools-tab-line"></span><a id="headers-tab" tabindex="-1" title="Headers" aria-controls="headers-panel" aria-selected="false" role="tab" data-tab-index="2">Headers</a></li></ul></nav><div class="panels"><div id="json-panel" style="visibility: visible; height: 100%;" class="tab-panel-box" role="tabpanel" aria-labelledby="json-tab"><div class="tab-panel json"><div class="jsonPanelBox tab-panel-inner"><div class="toolbar"><button class="btn save">Save</button><button class="btn copy">Copy</button><button class="btn collapse">Collapse All</button><button class="btn expand">Expand All</button><div class="devtools-separator"></div><div class="devtools-searchbox"><input class="searchBox devtools-filterinput" placeholder="Filter JSON" value=""><button class="devtools-searchinput-clear" hidden=""></button></div></div><div class="panelContent" id="json-scrolling-panel" tabindex="0"><table class="treeTable" role="tree" tabindex="0" aria-label="" cellpadding="0" cellspacing="0"><thead role="presentation"><tr class="" role="presentation"><td class="" role="presentation" id="default"></td><td class="" style="width: 100%;" role="presentation" id="value"></td></tr></thead><tbody role="presentation" tabindex="-1"><tr id="/_class" role="treeitem" aria-level="1" aria-selected="false" class="treeRow stringRow opened"><td class="treeLabelCell" style="--tree-label-cell-indent: 0px;" role="presentation"><span class="treeIcon open" role="presentation"></span><span class="treeLabel stringLabel" data-level="0">_class</span></td><td class="treeValueCell stringCell" role="presentation"><span><span class="objectBox objectBox-string">"hudson.security.WhoAmI"</span></span></td></tr><tr id="/anonymous" role="treeitem" aria-level="1" aria-selected="false" class="treeRow booleanRow opened"><td class="treeLabelCell" style="--tree-label-cell-indent: 0px;" role="presentation"><span class="treeIcon open" role="presentation"></span><span class="treeLabel booleanLabel" data-level="0">anonymous</span></td><td class="treeValueCell booleanCell" role="presentation"><span><span class="objectBox objectBox-number">false</span></span></td></tr><tr id="/authenticated" role="treeitem" aria-level="1" aria-selected="false" class="treeRow booleanRow opened"><td class="treeLabelCell" style="--tree-label-cell-indent: 0px;" role="presentation"><span class="treeIcon open" role="presentation"></span><span class="treeLabel booleanLabel" data-level="0">authenticated</span></td><td class="treeValueCell booleanCell" role="presentation"><span><span class="objectBox objectBox-number">true</span></span></td></tr><tr id="/authorities" role="treeitem" aria-level="1" aria-selected="false" aria-expanded="true" class="treeRow objectRow hasChildren opened"><td class="treeLabelCell" style="--tree-label-cell-indent: 0px;" role="presentation"><span class="treeIcon theme-twisty open" role="presentation"></span><span class="treeLabel objectLabel" data-level="0">authorities</span></td><td class="treeValueCell objectCell" role="presentation"><span></span></td></tr><tr id="/authorities/0" role="treeitem" aria-level="2" aria-selected="false" class="treeRow stringRow opened"><td class="treeLabelCell" style="--tree-label-cell-indent: 16px;" role="presentation"><span class="treeIcon open" role="presentation"></span><span class="treeLabel stringLabel" data-level="1">0</span></td><td class="treeValueCell stringCell" role="presentation"><span><span class="objectBox objectBox-string">"authenticated"</span></span></td></tr><tr id="/authorities/1" role="treeitem" aria-level="2" aria-selected="false" class="treeRow stringRow opened"><td class="treeLabelCell" style="--tree-label-cell-indent: 16px;" role="presentation"><span class="treeIcon open" role="presentation"></span><span class="treeLabel stringLabel" data-level="1">1</span></td><td class="treeValueCell stringCell" role="presentation"><span><span class="objectBox objectBox-string">"jenkinsAdmin"</span></span></td></tr><tr id="/authorities/2" role="treeitem" aria-level="2" aria-selected="false" class="treeRow stringRow opened"><td class="treeLabelCell" style="--tree-label-cell-indent: 16px;" role="presentation"><span class="treeIcon open" role="presentation"></span><span class="treeLabel stringLabel" data-level="1">2</span></td><td class="treeValueCell stringCell" role="presentation"><span><span class="objectBox objectBox-string">"default-roles-test-realm"</span></span></td></tr><tr id="/authorities/3" role="treeitem" aria-level="2" aria-selected="false" class="treeRow stringRow opened"><td class="treeLabelCell" style="--tree-label-cell-indent: 16px;" role="presentation"><span class="treeIcon open" role="presentation"></span><span class="treeLabel stringLabel" data-level="1">3</span></td><td class="treeValueCell stringCell" role="presentation"><span><span class="objectBox objectBox-string">"offline_access"</span></span></td></tr><tr id="/authorities/4" role="treeitem" aria-level="2" aria-selected="false" class="treeRow stringRow opened"><td class="treeLabelCell" style="--tree-label-cell-indent: 16px;" role="presentation"><span class="treeIcon open" role="presentation"></span><span class="treeLabel stringLabel" data-level="1">4</span></td><td class="treeValueCell stringCell" role="presentation"><span><span class="objectBox objectBox-string">"uma_authorization"</span></span></td></tr><tr id="/name" role="treeitem" aria-level="1" aria-selected="false" class="treeRow stringRow opened"><td class="treeLabelCell" style="--tree-label-cell-indent: 0px;" role="presentation"><span class="treeIcon open" role="presentation"></span><span class="treeLabel stringLabel" data-level="0">name</span></td><td class="treeValueCell stringCell" role="presentation"><span><span class="objectBox objectBox-string">"11d9bbf1-ded6-4023-a7ca-3a756c2850a2"</span></span></td></tr></tbody></table></div></div></div></div><div id="rawdata-panel" style="visibility: hidden; height: 0px;" class="tab-panel-box hidden" role="tabpanel" aria-labelledby="rawdata-tab"></div><div id="headers-panel" style="visibility: hidden; height: 0px;" class="tab-panel-box hidden" role="tabpanel" aria-labelledby="headers-tab"></div></div></div></div><script src="resource://devtools/client/jsonview/lib/require.js" data-main="resource://devtools/client/jsonview/viewer-config.js"></script></body></html>
```
while in Chrome
```
<html style="scroll-behavior: initial;"><head><meta name="color-scheme" content="light dark"><meta charset="utf-8"></head><body><pre>{"_class":"hudson.security.WhoAmI","anonymous":false,"authenticated":true,"authorities":["authenticated","jenkinsAdmin","default-roles-test-realm","offline_access","uma_authorization"],"name":"e860be16-d8b9-4fc2-9ca6-f1dcb65169c4"}</pre><div class="json-formatter-container"></div></body></html>
```

and `CapybaraPortingLayerImpl.getPageContent(this.driver) `will generate in firefox
```
JSON
Raw Data
Headers
Save
Copy
Collapse All
Expand All
_class "hudson.security.WhoAmI"
anonymous false
authenticated true
authorities
0 "authenticated"
1 "jenkinsAdmin"
2 "default-roles-test-realm"
3 "offline_access"
4 "uma_authorization"
name "11d9bbf1-ded6-4023-a7ca-3a756c2850a2"
```
as a result of 
![image](https://github.com/user-attachments/assets/b09709b5-a9ab-449d-ab35-c0ea4e3825f6)

while in Chrome
```
{"_class":"hudson.security.WhoAmI","anonymous":false,"authenticated":true,"authorities":["authenticated","jenkinsAdmin","default-roles-test-realm","offline_access","uma_authorization"],"name":"e860be16-d8b9-4fc2-9ca6-f1dcb65169c4"}
```
as a result of
![image](https://github.com/user-attachments/assets/e02c8c3f-7794-4c3e-9f61-5a612b9ea7bc)

I thought check the browser and process the output accordingly, but it seemed too dirty to me

Reply:
Then I tried the `api/xml`, but both `jenkins.getPageSource()` and `CapybaraPortingLayerImpl.getPageContent(this.driver)` fail as they are expecting to read HTML and this is not HTML but XML

* https://github.com/jenkinsci/acceptance-test-harness/blob/fda9804fe3e68cf63946ed747bcddd7966af4f2f/src/main/java/org/jenkinsci/test/acceptance/po/CapybaraPortingLayerImpl.java#L531-L539
* https://github.com/jenkinsci/acceptance-test-harness/blob/fda9804fe3e68cf63946ed747bcddd7966af4f2f/src/main/java/org/jenkinsci/test/acceptance/po/CapybaraPortingLayerImpl.java#L541-L546

Reply:
So I opted for the UI, as it seems to me the cleanest solution

Reply:
for self education you can check out [`getJson`](https://github.com/jenkinsci/acceptance-test-harness/blob/fda9804fe3e68cf63946ed747bcddd7966af4f2f/src/main/java/org/jenkinsci/test/acceptance/po/ContainerPageObject.java#L56-L79) to see how it solves this :) 

to get the page, you can interact with webdriver directly (`driver.getPageSource()`).  probably need to turn off some pretty printing, or to use the `view-source:URL` style.

Anyway, you can leave this as is, if/when it breaks we can improve it

Reply:
let me check `getJson`. I didn't find that method and if it works, I prefer to use the JSON object

`driver.getPageSource()` returns the same output as `jenkins.getPageSource()`, I checked that as well

Reply:
ah! I didn't find that method because I looked for something in the `CapybaraPortingLayerImpl` class